File: app/src/main/java/me/zhanghai/android/files/provider/archive/LocalArchiveFileSystemProvider.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.archive;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java8.nio.channels.FileChannel;
import java8.nio.channels.SeekableByteChannel;
import java8.nio.file.AccessDeniedException;
import java8.nio.file.AccessMode;
import java8.nio.file.CopyOption;
import java8.nio.file.DirectoryStream;
import java8.nio.file.FileStore;
import java8.nio.file.FileSystem;
import java8.nio.file.Files;
import java8.nio.file.LinkOption;
import java8.nio.file.OpenOption;
import java8.nio.file.Path;
import java8.nio.file.Paths;
import java8.nio.file.ProviderMismatchException;
import java8.nio.file.attribute.BasicFileAttributes;
import java8.nio.file.attribute.FileAttribute;
import java8.nio.file.attribute.FileAttributeView;
import java8.nio.file.spi.FileSystemProvider;
import java9.util.function.Consumer;
import me.zhanghai.android.files.provider.common.AccessModes;
import me.zhanghai.android.files.provider.common.ByteString;
import me.zhanghai.android.files.provider.common.ByteStringPath;
import me.zhanghai.android.files.provider.common.ByteStringUriUtils;
import me.zhanghai.android.files.provider.common.FileSystemCache;
import me.zhanghai.android.files.provider.common.OpenOptions;
import me.zhanghai.android.files.provider.common.PathListDirectoryStream;
import me.zhanghai.android.files.provider.common.ReadOnlyFileSystemException;
import me.zhanghai.android.files.provider.common.Searchable;
import me.zhanghai.android.files.provider.common.WalkFileTreeSearchable;

class LocalArchiveFileSystemProvider extends FileSystemProvider implements Searchable {

    static final String SCHEME = "archive";

    private ArchiveFileSystemProvider mProvider;

    @NonNull
    private final FileSystemCache<Path, ArchiveFileSystem> mFileSystems = new FileSystemCache<>();

    LocalArchiveFileSystemProvider(@NonNull ArchiveFileSystemProvider provider) {
        mProvider = provider;
    }

    static boolean isArchivePath(@NonNull Path path) {
        Objects.requireNonNull(path);
        return path instanceof ArchivePath;
    }

    @NonNull
    static Path getArchiveFile(@NonNull Path path) {
        requireArchivePath(path);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) path.getFileSystem();
        return fileSystem.getArchiveFile();
    }

    static void refresh(@NonNull Path path) {
        requireArchivePath(path);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) path.getFileSystem();
        fileSystem.refresh();
    }

    @NonNull
    Path getRootPathForArchiveFile(@NonNull Path archiveFile) {
        return getOrNewFileSystem(archiveFile).getRootDirectory();
    }

    @NonNull
    ArchiveFileSystem getOrNewFileSystem(@NonNull Path archiveFile) {
        Objects.requireNonNull(archiveFile);
        return mFileSystems.getOrNew(archiveFile, () -> newFileSystem(archiveFile));
    }

    @NonNull
    @Override
    public String getScheme() {
        return SCHEME;
    }

    @NonNull
    @Override
    public FileSystem newFileSystem(@NonNull URI uri, @NonNull Map<String, ?> env) {
        Objects.requireNonNull(uri);
        requireSameScheme(uri);
        Objects.requireNonNull(env);
        Path archiveFile = getArchiveFileFromUri(uri);
        return mFileSystems.new_(archiveFile, () -> newFileSystem(archiveFile));
    }

    @NonNull
    @Override
    public ArchiveFileSystem getFileSystem(@NonNull URI uri) {
        Objects.requireNonNull(uri);
        requireSameScheme(uri);
        Path archiveFile = getArchiveFileFromUri(uri);
        return mFileSystems.get(archiveFile);
    }

    @NonNull
    private static Path getArchiveFileFromUri(@NonNull URI uri) {
        ByteString schemeSpecificPart = ByteStringUriUtils.getDecodedSchemeSpecificPart(uri);
        if (schemeSpecificPart == null) {
            throw new IllegalArgumentException("URI must have a scheme specific part");
        }
        URI archiveUri;
        try {
            archiveUri = new URI(schemeSpecificPart.toString());
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException(e);
        }
        return Paths.get(archiveUri);
    }

    void removeFileSystem(@NonNull ArchiveFileSystem fileSystem) {
        mFileSystems.remove(fileSystem.getArchiveFile(), fileSystem);
    }

    @NonNull
    @Override
    public Path getPath(@NonNull URI uri) {
        Objects.requireNonNull(uri);
        requireSameScheme(uri);
        Path archiveFile = getArchiveFileFromUri(uri);
        ByteString fragment = ByteStringUriUtils.getDecodedFragment(uri);
        if (fragment == null) {
            throw new IllegalArgumentException("URI must have a fragment");
        }
        return getOrNewFileSystem(archiveFile).getPath(fragment);
    }

    private static void requireSameScheme(@NonNull URI uri) {
        if (!Objects.equals(uri.getScheme(), SCHEME)) {
            throw new IllegalArgumentException("URI scheme must be \"" + SCHEME + "\"");
        }
    }

    @NonNull
    @Override
    public FileSystem newFileSystem(@NonNull Path file, @NonNull Map<String, ?> env) {
        Objects.requireNonNull(file);
        Objects.requireNonNull(env);
        return newFileSystem(file);
    }

    @NonNull
    private ArchiveFileSystem newFileSystem(@NonNull Path archiveFile) {
        return new ArchiveFileSystem(mProvider, archiveFile);
    }

    @NonNull
    @Override
    public InputStream newInputStream(@NonNull Path file, @NonNull OpenOption... options)
            throws IOException {
        requireArchivePath(file);
        Objects.requireNonNull(options);
        OpenOptions openOptions = OpenOptions.fromArray(options);
        ArchiveOpenOptions.check(openOptions);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) file.getFileSystem();
        return fileSystem.newInputStreamAsLocal(file);
    }

    @NonNull
    @Override
    public FileChannel newFileChannel(@NonNull Path file,
                                      @NonNull Set<? extends OpenOption> options,
                                      @NonNull FileAttribute<?>... attributes) {
        requireArchivePath(file);
        Objects.requireNonNull(options);
        Objects.requireNonNull(attributes);
        OpenOptions openOptions = OpenOptions.fromSet(options);
        ArchiveOpenOptions.check(openOptions);
        if (attributes.length > 0) {
            throw new UnsupportedOperationException(Arrays.toString(attributes));
        }
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    public SeekableByteChannel newByteChannel(@NonNull Path file,
                                              @NonNull Set<? extends OpenOption> options,
                                              @NonNull FileAttribute<?>... attributes) {
        requireArchivePath(file);
        Objects.requireNonNull(options);
        Objects.requireNonNull(attributes);
        OpenOptions openOptions = OpenOptions.fromSet(options);
        ArchiveOpenOptions.check(openOptions);
        if (attributes.length > 0) {
            throw new UnsupportedOperationException(Arrays.toString(attributes));
        }
        throw new UnsupportedOperationException();
    }

    @NonNull
    @Override
    public DirectoryStream<Path> newDirectoryStream(
            @NonNull Path directory, @NonNull DirectoryStream.Filter<? super Path> filter)
            throws IOException {
        requireArchivePath(directory);
        Objects.requireNonNull(filter);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) directory.getFileSystem();
        List<Path> children = fileSystem.getDirectoryChildrenAsLocal(directory);
        return new PathListDirectoryStream(children, filter);
    }

    @Override
    public void createDirectory(@NonNull Path directory, @NonNull FileAttribute<?>... attributes)
            throws IOException{
        requireArchivePath(directory);
        Objects.requireNonNull(attributes);
        throw new ReadOnlyFileSystemException(directory.toString());
    }

    @Override
    public void createSymbolicLink(@NonNull Path link, @NonNull Path target,
                                   @NonNull FileAttribute<?>... attributes) throws IOException {
        requireArchivePath(link);
        requireArchiveOrByteStringPath(target);
        Objects.requireNonNull(attributes);
        throw new ReadOnlyFileSystemException(link.toString(), target.toString(), null);
    }

    @Override
    public void createLink(@NonNull Path link, @NonNull Path existing) throws IOException {
        requireArchivePath(link);
        requireArchivePath(existing);
        throw new ReadOnlyFileSystemException(link.toString(), existing.toString(), null);
    }

    @Override
    public void delete(@NonNull Path path) throws IOException {
        requireArchivePath(path);
        throw new ReadOnlyFileSystemException(path.toString());
    }

    @NonNull
    @Override
    public Path readSymbolicLink(@NonNull Path link) throws IOException {
        requireArchivePath(link);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) link.getFileSystem();
        String target = fileSystem.readSymbolicLinkAsLocal(link);
        return new ByteStringPath(ByteString.fromString(target));
    }

    @Override
    public void copy(@NonNull Path source, @NonNull Path target, @NonNull CopyOption... options)
            throws IOException {
        requireArchivePath(source);
        requireArchivePath(target);
        Objects.requireNonNull(options);
        throw new ReadOnlyFileSystemException(source.toString(), target.toString(), null);
    }

    @Override
    public void move(@NonNull Path source, @NonNull Path target, @NonNull CopyOption... options)
            throws IOException {
        requireArchivePath(source);
        requireArchivePath(target);
        Objects.requireNonNull(options);
        throw new ReadOnlyFileSystemException(source.toString(), target.toString(), null);
    }

    @Override
    public boolean isSameFile(@NonNull Path path, @NonNull Path path2) throws IOException {
        requireArchivePath(path);
        Objects.requireNonNull(path2);
        if (Objects.equals(path, path2)) {
            return true;
        }
        if (!(path2 instanceof ArchivePath)) {
            return false;
        }
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) path.getFileSystem();
        ArchiveFileSystem fileSystem2 = (ArchiveFileSystem) path2.getFileSystem();
        if (!Files.isSameFile(fileSystem.getArchiveFile(), fileSystem2.getArchiveFile())) {
            return false;
        }
        return Objects.equals(path, fileSystem.getPath(path2.toString()));
    }

    @Override
    public boolean isHidden(@NonNull Path path) {
        requireArchivePath(path);
        return false;
    }

    @NonNull
    @Override
    public FileStore getFileStore(@NonNull Path path) {
        requireArchivePath(path);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) path.getFileSystem();
        Path archiveFile = fileSystem.getArchiveFile();
        return new ArchiveFileStore(archiveFile);
    }

    @Override
    public void checkAccess(@NonNull Path path, @NonNull AccessMode... modes) throws IOException {
        requireArchivePath(path);
        Objects.requireNonNull(modes);
        AccessModes accessModes = AccessModes.fromArray(modes);
        ArchiveFileSystem fileSystem = (ArchiveFileSystem) path.getFileSystem();
        fileSystem.getEntryAsLocal(path);
        if (accessModes.hasWrite() || accessModes.hasExecute()) {
            throw new AccessDeniedException(path.toString());
        }
    }

    @Nullable
    @Override
    public <V extends FileAttributeView> V getFileAttributeView(@NonNull Path path,
                                                                @NonNull Class<V> type,
                                                                @NonNull LinkOption... options) {
        requireArchivePath(path);
        Objects.requireNonNull(type);
        Objects.requireNonNull(options);
        if (!supportsFileAttributeView(type)) {
            return null;
        }
        //noinspection unchecked
        return (V) getFileAttributeView(path);
    }

    static boolean supportsFileAttributeView(@NonNull Class<? extends FileAttributeView> type) {
        return type.isAssignableFrom(ArchiveFileAttributeView.class);
    }

    @NonNull
    @Override
    public <A extends BasicFileAttributes> A readAttributes(@NonNull Path path,
                                                            @NonNull Class<A> type,
                                                            @NonNull LinkOption... options)
            throws IOException {
        requireArchivePath(path);
        Objects.requireNonNull(type);
        Objects.requireNonNull(options);
        if (!type.isAssignableFrom(ArchiveFileAttributes.class)) {
            throw new UnsupportedOperationException(type.toString());
        }
        //noinspection unchecked
        return (A) getFileAttributeView(path).readAttributes();
    }

    private static ArchiveFileAttributeView getFileAttributeView(@NonNull Path path) {
        return new ArchiveFileAttributeView(path);
    }

    @NonNull
    @Override
    public Map<String, Object> readAttributes(@NonNull Path path, @NonNull String attributes,
                                              @NonNull LinkOption... options) {
        requireArchivePath(path);
        Objects.requireNonNull(attributes);
        Objects.requireNonNull(options);
        throw new UnsupportedOperationException();
    }

    @Override
    public void setAttribute(@NonNull Path path, @NonNull String attribute, @NonNull Object value,
                             @NonNull LinkOption... options) {
        requireArchivePath(path);
        Objects.requireNonNull(attribute);
        Objects.requireNonNull(value);
        Objects.requireNonNull(options);
        throw new UnsupportedOperationException();
    }

    @Override
    public void search(@NonNull Path directory, @NonNull String query,
                       @NonNull Consumer<List<Path>> listener, long intervalMillis)
            throws IOException {
        requireArchivePath(directory);
        Objects.requireNonNull(query);
        Objects.requireNonNull(listener);
        WalkFileTreeSearchable.search(directory, query, listener, intervalMillis);
    }

    private static void requireArchivePath(@NonNull Path path) {
        Objects.requireNonNull(path);
        if (!(path instanceof ArchivePath)) {
            throw new ProviderMismatchException(path.toString());
        }
    }

    private static ByteString requireArchiveOrByteStringPath(@NonNull Path path) {
        Objects.requireNonNull(path);
        if (path instanceof ArchivePath) {
            return ((ArchivePath) path).toByteString();
        } else if (path instanceof ByteStringPath) {
            return ((ByteStringPath) path).toByteString();
        } else {
            throw new ProviderMismatchException(path.toString());
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/archive/LocalArchiveFileSystemProvider.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.archive

import java8.nio.channels.FileChannel
import java8.nio.channels.SeekableByteChannel
import java8.nio.file.AccessDeniedException
import java8.nio.file.AccessMode
import java8.nio.file.CopyOption
import java8.nio.file.DirectoryStream
import java8.nio.file.FileStore
import java8.nio.file.FileSystem
import java8.nio.file.LinkOption
import java8.nio.file.OpenOption
import java8.nio.file.Path
import java8.nio.file.Paths
import java8.nio.file.ProviderMismatchException
import java8.nio.file.attribute.BasicFileAttributes
import java8.nio.file.attribute.FileAttribute
import java8.nio.file.attribute.FileAttributeView
import java8.nio.file.spi.FileSystemProvider
import me.zhanghai.android.files.provider.common.ByteStringPath
import me.zhanghai.android.files.provider.common.FileSystemCache
import me.zhanghai.android.files.provider.common.PathListDirectoryStream
import me.zhanghai.android.files.provider.common.ReadOnlyFileSystemException
import me.zhanghai.android.files.provider.common.Searchable
import me.zhanghai.android.files.provider.common.WalkFileTreeSearchable
import me.zhanghai.android.files.provider.common.decodedFragmentByteString
import me.zhanghai.android.files.provider.common.decodedSchemeSpecificPartByteString
import me.zhanghai.android.files.provider.common.isSameFile
import me.zhanghai.android.files.provider.common.toAccessModes
import me.zhanghai.android.files.provider.common.toByteString
import me.zhanghai.android.files.provider.common.toOpenOptions
import java.io.IOException
import java.io.InputStream
import java.net.URI

class LocalArchiveFileSystemProvider(
    private val provider: ArchiveFileSystemProvider
) : FileSystemProvider(), Searchable {
    private val fileSystems = FileSystemCache<Path, ArchiveFileSystem>()

    override fun getScheme(): String = SCHEME

    override fun newFileSystem(uri: URI, env: Map<String, *>): FileSystem {
        uri.requireSameScheme()
        val archiveFile = uri.archiveFile
        return fileSystems.create(archiveFile) { newFileSystem(archiveFile) }
    }

    override fun newFileSystem(file: Path, env: Map<String, *>): FileSystem = newFileSystem(file)

    internal fun getOrNewFileSystem(archiveFile: Path): ArchiveFileSystem =
        fileSystems.getOrCreate(archiveFile) { newFileSystem(archiveFile) }

    private fun newFileSystem(archiveFile: Path): ArchiveFileSystem =
        ArchiveFileSystem(provider, archiveFile)

    override fun getFileSystem(uri: URI): FileSystem {
        uri.requireSameScheme()
        val archiveFile = uri.archiveFile
        return fileSystems[archiveFile]
    }

    internal fun removeFileSystem(fileSystem: ArchiveFileSystem) {
        fileSystems.remove(fileSystem.archiveFile, fileSystem)
    }

    override fun getPath(uri: URI): Path {
        uri.requireSameScheme()
        val archiveFile = uri.archiveFile
        val fragment = uri.decodedFragmentByteString
            ?: throw IllegalArgumentException("URI must have a fragment")
        return getOrNewFileSystem(archiveFile).getPath(fragment)
    }

    private fun URI.requireSameScheme() {
        val scheme = scheme
        require(scheme == SCHEME) { "URI scheme $scheme must be $SCHEME" }
    }

    private val URI.archiveFile: Path
        get() {
            val schemeSpecificPart = decodedSchemeSpecificPartByteString
                ?: throw IllegalArgumentException("URI must have a scheme specific part")
            val archiveUri = URI.create(schemeSpecificPart.toString())
            return Paths.get(archiveUri)
        }

    @Throws(IOException::class)
    override fun newInputStream(file: Path, vararg options: OpenOption): InputStream {
        file as? ArchivePath ?: throw ProviderMismatchException(file.toString())
        options.toOpenOptions().checkForArchive()
        return file.fileSystem.newInputStreamAsLocal(file)
    }

    override fun newFileChannel(
        file: Path,
        options: Set<OpenOption>,
        vararg attributes: FileAttribute<*>
    ): FileChannel {
        file as? ArchivePath ?: throw ProviderMismatchException(file.toString())
        options.toOpenOptions().checkForArchive()
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(attributes.contentToString())
        }
        throw UnsupportedOperationException()
    }

    override fun newByteChannel(
        file: Path,
        options: Set<OpenOption>,
        vararg attributes: FileAttribute<*>
    ): SeekableByteChannel {
        file as? ArchivePath ?: throw ProviderMismatchException(file.toString())
        options.toOpenOptions().checkForArchive()
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(attributes.contentToString())
        }
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun newDirectoryStream(
        directory: Path,
        filter: DirectoryStream.Filter<in Path>
    ): DirectoryStream<Path> {
        directory as? ArchivePath ?: throw ProviderMismatchException(directory.toString())
        val children = directory.fileSystem.getDirectoryChildrenAsLocal(directory)
        return PathListDirectoryStream(children, filter)
    }

    @Throws(IOException::class)
    override fun createDirectory(directory: Path, vararg attributes: FileAttribute<*>) {
        directory as? ArchivePath ?: throw ProviderMismatchException(directory.toString())
        throw ReadOnlyFileSystemException(directory.toString())
    }

    @Throws(IOException::class)
    override fun createSymbolicLink(link: Path, target: Path, vararg attributes: FileAttribute<*>) {
        link as? ArchivePath ?: throw ProviderMismatchException(link.toString())
        when (target) {
            is ArchivePath, is ByteStringPath -> {}
            else -> throw ProviderMismatchException(target.toString())
        }
        throw ReadOnlyFileSystemException(link.toString(), target.toString(), null)
    }

    @Throws(IOException::class)
    override fun createLink(link: Path, existing: Path) {
        link as? ArchivePath ?: throw ProviderMismatchException(link.toString())
        existing as? ArchivePath ?: throw ProviderMismatchException(existing.toString())
        throw ReadOnlyFileSystemException(link.toString(), existing.toString(), null)
    }

    @Throws(IOException::class)
    override fun delete(path: Path) {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        throw ReadOnlyFileSystemException(path.toString())
    }

    @Throws(IOException::class)
    override fun readSymbolicLink(link: Path): Path {
        link as? ArchivePath ?: throw ProviderMismatchException(link.toString())
        val target = link.fileSystem.readSymbolicLinkAsLocal(link)
        return ByteStringPath(target.toByteString())
    }

    @Throws(IOException::class)
    override fun copy(source: Path, target: Path, vararg options: CopyOption) {
        source as? ArchivePath ?: throw ProviderMismatchException(source.toString())
        target as? ArchivePath ?: throw ProviderMismatchException(target.toString())
        throw ReadOnlyFileSystemException(source.toString(), target.toString(), null)
    }

    @Throws(IOException::class)
    override fun move(source: Path, target: Path, vararg options: CopyOption) {
        source as? ArchivePath ?: throw ProviderMismatchException(source.toString())
        target as? ArchivePath ?: throw ProviderMismatchException(target.toString())
        throw ReadOnlyFileSystemException(source.toString(), target.toString(), null)
    }

    @Throws(IOException::class)
    override fun isSameFile(path: Path, path2: Path): Boolean {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        if (path == path2) {
            return true
        }
        if (path2 !is ArchivePath) {
            return false
        }
        val fileSystem = path.fileSystem
        if (!fileSystem.archiveFile.isSameFile(path2.fileSystem.archiveFile)) {
            return false
        }
        return path == fileSystem.getPath(path2.toString())
    }

    override fun isHidden(path: Path): Boolean {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        return false
    }

    override fun getFileStore(path: Path): FileStore {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        val archiveFile = path.fileSystem.archiveFile
        return ArchiveFileStore(archiveFile)
    }

    @Throws(IOException::class)
    override fun checkAccess(path: Path, vararg modes: AccessMode) {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        val accessModes = modes.toAccessModes()
        path.fileSystem.getEntryAsLocal(path)
        if (accessModes.write || accessModes.execute) {
            throw AccessDeniedException(path.toString())
        }
    }

    override fun <V : FileAttributeView> getFileAttributeView(
        path: Path,
        type: Class<V>,
        vararg options: LinkOption
    ): V? {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        if (!supportsFileAttributeView(type)) {
            return null
        }
        @Suppress("UNCHECKED_CAST")
        return getFileAttributeView(path) as V
    }

    @Throws(IOException::class)
    override fun <A : BasicFileAttributes> readAttributes(
        path: Path,
        type: Class<A>,
        vararg options: LinkOption
    ): A {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        if (!type.isAssignableFrom(ArchiveFileAttributes::class.java)) {
            throw UnsupportedOperationException(type.toString())
        }
        @Suppress("UNCHECKED_CAST")
        return getFileAttributeView(path).readAttributes() as A
    }

    private fun getFileAttributeView(path: ArchivePath): ArchiveFileAttributeView =
        ArchiveFileAttributeView(path)

    override fun readAttributes(
        path: Path,
        attributes: String,
        vararg options: LinkOption
    ): Map<String, Any> {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        throw UnsupportedOperationException()
    }

    override fun setAttribute(
        path: Path,
        attribute: String,
        value: Any,
        vararg options: LinkOption
    ) {
        path as? ArchivePath ?: throw ProviderMismatchException(path.toString())
        throw UnsupportedOperationException()
    }

    @Throws(IOException::class)
    override fun search(
        directory: Path,
        query: String,
        intervalMillis: Long,
        listener: (List<Path>) -> Unit
    ) {
        directory as? ArchivePath ?: throw ProviderMismatchException(directory.toString())
        WalkFileTreeSearchable.search(directory, query, intervalMillis, listener)
    }

    companion object {
        internal const val SCHEME = "archive"

        internal fun supportsFileAttributeView(type: Class<out FileAttributeView>): Boolean =
            type.isAssignableFrom(ArchiveFileAttributeView::class.java)
    }
}
